lib/prometheus_exporter/instrumentation/unicorn.rb

```diff
@@ -0,0 +1,68 @@
+begin
+  require 'raindrops'
+rescue LoadError
+  # No raindrops available, dont do anything
+end
+
+module PrometheusExporter::Instrumentation
+  # collects stats from unicorn
+  class Unicorn
+    def self.start(pid_file:, socket:, client:, frequency: 5)
+      unicorn_collector = new(pid_file: pid_file, socket: socket)
+      client ||= PrometheusExporter::Client.default
+      Thread.new do
+        loop do
+          metric = unicorn_collector.collect
+          client.send_json metric
+        rescue StandardError => e
+          STDERR.puts("Prometheus Exporter Failed To Collect Unicorn Stats #{e}")
+        ensure
+          sleep frequency
+        end
+      end
+    end
+
+    def initialize(pid_file:, socket:)
+      @pid_file = pid_file
+      @socket = socket
+      @tcp = socket =~ /\A.+:\d+\z/
+    end
+
+    def collect
+      metric = {}
+      metric[:type] = 'unicorn'
+      collect_unicorn_stats(metric)
+      metric
+    end
+
+    def collect_unicorn_stats(metric)
+      stats = socket_stats
+
+      metric[:active_workers_total] = stats.active
+      metric[:request_backlog_total] = stats.queued
+      metric[:workers_total] = worker_process_count
+    end
+
+    private
+
+    def worker_process_count
+      return nil unless File.exist?(@pid_file)
+      pid = File.read(@pid_file)
+
+      return nil unless pid && pid.to_i > 0
+
+      # find all processes whose parent is the unicorn master
+      # but we're actually only interested in the number of processes (= lines of output)
+      result = `ps --no-header -o pid --ppid #{pid}`
+      result.lines.count
+    end
+
+    def socket_stats
+      if @tcp
+        Raindrops::Linux.tcp_listener_stats(@socket)[@socket]
```

Comment:
This can take either an address or a socket per: https://www.rubydoc.info/gems/raindrops/0.13.0/Raindrops%2FLinux.tcp_listener_stats so we should clean up the params here a bit.

Reply:
Sorry, I dont really get what you mean here... :p

Reply:
sorry I mean this is legit: (in fact it is what we do in discourse) 

```
Raindrops::Linux.tcp_listener_stats(['127.0.0.1:3000'])['127.0.0.1:3000']
```

So maybe add an extra param to the collector so this is clear `--unicorn-address http://localhost:3000`

Reply:
`@socket` is that address... I'd rename that param to `listener_address` to be clear?

Reply:
yeah I guess that works ... was thinking maybe people want to be explicit if they use a `unicorn.sock` sock file here, but address works I guess.